src/main/java/com/gradle/CustomBuildScanEnhancements.java

```diff
@@ -396,19 +397,21 @@ public void execute(BuildScanExtension buildScan) {
             }
 
             if (isNotEmpty(gitRepo) && isNotEmpty(gitCommitId)) {
-                if (gitRepo.contains("github.com/") || gitRepo.contains("github.com:")) {
-                    Matcher matcher = Pattern.compile("(.*)github\\.com[/|:](.*)").matcher(gitRepo);
-                    if (matcher.matches()) {
-                        String rawRepoPath = matcher.group(2);
-                        String repoPath = rawRepoPath.endsWith(".git") ? rawRepoPath.substring(0, rawRepoPath.length() - 4) : rawRepoPath;
-                        buildScan.link("Github source", "https://github.com/" + repoPath + "/tree/" + gitCommitId);
-                    }
-                } else if (gitRepo.contains("gitlab.com/") || gitRepo.contains("gitlab.com:")) {
-                    Matcher matcher = Pattern.compile("(.*)gitlab\\.com[/|:](.*)").matcher(gitRepo);
+                Optional<String> gitHubUrl = envVariable("GITHUB_SERVER_URL", providers);
+                Optional<String> gitHubRepository = envVariable("GITHUB_REPOSITORY", providers);
+                if (gitHubUrl.isPresent() && gitHubRepository.isPresent()) {
+                    buildScan.link("GitHub source", gitHubUrl + "/" + gitHubRepository + "/tree/" + gitCommitId);
+                } else {
+                    Matcher matcher = Pattern.compile("(https://|.*?@)(.*?(github|gitlab).*?)[:/](.*)").matcher(gitRepo);
                     if (matcher.matches()) {
-                        String rawRepoPath = matcher.group(2);
+                        String repoUrl = "https://" + matcher.group(2) + "/";
+                        String rawRepoPath = matcher.group(4);
                         String repoPath = rawRepoPath.endsWith(".git") ? rawRepoPath.substring(0, rawRepoPath.length() - 4) : rawRepoPath;
```

Comment:
I think we could remove this logic that strips `.git` by adding it as a group to the end of the regular expression above: `...[:/](.*)(?:\.git)`. Then the regular expression would completely capture all of the string manipulation we are doing here.

Reply:
Agreed, that should clean things up a bit.

Addressed in 4a5f2ca